File: sysplant/templates/iterators/freshy.nim
# Parse Export Directory and lookup syscall by name (start with Nt and not Ntdll), sort addresses to retrieve syscall number https://github.com/crummie5/FreshyCalls
iterator SPT_Iterator(mi: MODULEINFO): (int32, PBYTE) =
    # Extract headers
    let codeBase = mi.lpBaseOfDll
    let dosHeader = cast[PIMAGE_DOS_HEADER](codeBase)
    let ntHeader = cast[PIMAGE_NT_HEADERS](cast[DWORD_PTR](codeBase) + dosHeader.e_lfanew)
    
    # Get export table
    let directory = ntHeader.OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_EXPORT]
    let exports = codeBase{directory.VirtualAddress}[PIMAGE_EXPORT_DIRECTORY]
    
    var
        nameRef = codeBase{exports.AddressOfNames}[PDWORD]
        funcRef = codeBase{exports.AddressOfFunctions}[PDWORD]
        ordinal = codeBase{exports.AddressOfNameOrdinals}[PWORD]
    
    # Search Begin Address in Export Table
    for j in 0 ..< exports.NumberOfFunctions:
        let
            name = $(codeBase{nameRef[]}[LPCSTR])
            offset = funcRef[ordinal[j][int]]
            address = codeBase{offset}[PBYTE]

        # Check offset with current function, ensure this is a syscall
        if name.startsWith("Nt") and not name.startsWith("Ntdll"):
            let hash: int32 = SPT_HashSyscallName(name)
            # Calculate jmp address avoiding EDR hooks
            yield (hash, address)
            
        ++nameRef

proc SPT_PopulateSyscalls =
    # Return short when work is already done
    if len(ssdt) > 0:
        return
    
    # Could set arg parse to check different process
    let pid = GetCurrentProcessId()

    # Create module snapshot
    let hModule = CreateToolhelp32Snapshot(TH32CS_SNAPMODULE, pid)
    defer: CloseHandle(hModule)

    # Store process handle
    let handle = OpenProcess(PROCESS_ALL_ACCESS, FALSE, pid)

    type Entry = tuple
        address: PBYTE
        hash: int32
    
    var
        me32: MODULEENTRY32
        mi: MODULEINFO
        tmp: seq[Entry]
    
    me32.dwSize = cast[DWORD](sizeof(MODULEENTRY32))

    # SKip to ntdll.dll
    hModule.Module32First(addr me32)
    # NTDLL.dll is not guaranteed to be first
    while hModule.Module32Next(addr me32):
        # Detect \ntdll.dll
        if "92110116100108108461001081080" in me32.szExePath.join():
            break
    
    handle.GetModuleInformation(me32.hModule, addr mi, cast[DWORD](sizeof(mi)))

    # Resolve address
    for hash, address in mi.SPT_Iterator():
        tmp.add((address, hash))
    
    # Sort syscalls by address
    tmp.sort(system.cmp)

    var padding = 0x0
    # Register syscalls
    for i in 0 ..< len(tmp):
        let
            address = tmp[i][0]
            hash = tmp[i][1]
        
        # All syscall stub are identical for a Windows version
        if padding == 0x0:
            padding = SPT_DetectPadding(address)

        ssdt[hash] = Syscall(address: address[PVOID], ssn: i[int32], syscallAddress: address{padding}[PVOID])

    return

File: sysplant/templates/iterators/freshy.rs
// Parse Export Directory and lookup syscall by name (start with Nt and not Ntdll),
// sort addresses to retrieve syscall number
// https://github.com/crummie5/FreshyCalls
#[unsafe(no_mangle)]
unsafe fn SPT_PopulateSyscallList() -> BOOL {
    if SPT_SyscallList.Count != 0 {
        return TRUE;
    }

    let peb = get_peb();
    let ldr = (*peb).Ldr;
    let mut export_directory: *mut IMAGE_EXPORT_DIRECTORY = ptr::null_mut();
    let mut dll_base: PVOID = ptr::null_mut();

    // Get the DllBase address of NTDLL.dll.
    let mut ldr_entry = (*ldr).Reserved2[1] as *mut SPT_LDR_DATA_TABLE_ENTRY;
    while !(*ldr_entry).DllBase.is_null() {
        dll_base = (*ldr_entry).DllBase;
        let dos_header = dll_base as *const IMAGE_DOS_HEADER;
        let nt_headers = SPT_RVA2VA!(*const IMAGE_NT_HEADERS64, dll_base, (*dos_header).e_lfanew);
        let data_directory = (*nt_headers).OptionalHeader.DataDirectory.as_ptr();
        let va = (*data_directory.add(IMAGE_DIRECTORY_ENTRY_EXPORT)).VirtualAddress;
        if va == 0 {
            ldr_entry = (*ldr_entry).Reserved1[0] as *mut SPT_LDR_DATA_TABLE_ENTRY;
            continue;
        }

        export_directory = SPT_RVA2VA!(*mut IMAGE_EXPORT_DIRECTORY, dll_base, va);

        let dll_name = SPT_RVA2VA!(*const u8, dll_base, (*export_directory).Name);
        if (*(dll_name as *const u32) | 0x20202020) != 0x6c64746e {
            ldr_entry = (*ldr_entry).Reserved1[0] as *mut SPT_LDR_DATA_TABLE_ENTRY;
            continue;
        }
        if (*(dll_name.add(4) as *const u32) | 0x20202020) == 0x6c642e6c {
            break;
        }
        ldr_entry = (*ldr_entry).Reserved1[0] as *mut SPT_LDR_DATA_TABLE_ENTRY;
    }

    if export_directory.is_null() {
        return FALSE;
    }

    let mut number_of_names = (*export_directory).NumberOfNames;
    let functions = SPT_RVA2VA!(*const DWORD, dll_base, (*export_directory).AddressOfFunctions);
    let names = SPT_RVA2VA!(*const DWORD, dll_base, (*export_directory).AddressOfNames);
    let ordinals = SPT_RVA2VA!(*const WORD, dll_base, (*export_directory).AddressOfNameOrdinals);

    // Populate SPT_SyscallList with unsorted Zw* entries.
    let mut i: u32 = 0;
    let mut padding: u16 = 0;
    let entries = SPT_SyscallList.Entries.as_mut_ptr();
    while number_of_names > 0 {
        number_of_names -= 1;
        let function_name = SPT_RVA2VA!(*const i8, dll_base, *names.add(number_of_names as usize));
        let function_address = SPT_RVA2VA!(PVOID, dll_base, *functions.add(*ordinals.add(number_of_names as usize) as usize));

        // Is this a system call? "Zw" == 0x775a
        if *(function_name as *const u16) == 0x775a {
            (*entries.add(i as usize)).Hash = SPT_HashSyscallName(function_name);
            (*entries.add(i as usize)).Address = function_address;

            if padding == 0 {
                padding = SPT_DetectPadding((*entries.add(i as usize)).Address);
            }

            (*entries.add(i as usize)).SyscallAddress =
                SPT_RVA2VA!(PVOID, (*entries.add(i as usize)).Address, padding);

            i += 1;
            if i == SPT_MAX_ENTRIES as u32 {
                break;
            }
        }
    }

    // Save total number of system calls found.
    SPT_SyscallList.Count = i;

    // Sort the list by address in ascending order (bubble sort).
    for i in 0..(SPT_SyscallList.Count as usize).saturating_sub(1) {
        for j in 0..(SPT_SyscallList.Count as usize - i - 1) {
            if (*entries.add(j)).Address as usize > (*entries.add(j + 1)).Address as usize {
                // Swap entries
                let temp_hash = (*entries.add(j)).Hash;
                let temp_addr = (*entries.add(j)).Address;
                let temp_sys = (*entries.add(j)).SyscallAddress;

                (*entries.add(j)).Hash = (*entries.add(j + 1)).Hash;
                (*entries.add(j)).Address = (*entries.add(j + 1)).Address;
                (*entries.add(j)).SyscallAddress = (*entries.add(j + 1)).SyscallAddress;

                (*entries.add(j + 1)).Hash = temp_hash;
                (*entries.add(j + 1)).Address = temp_addr;
                (*entries.add(j + 1)).SyscallAddress = temp_sys;
            }
        }
    }

    TRUE
}
